### PR TITLE
Fixed double-counting of periopdic image pairs

### PR DESCRIPTION
**Pull request by Luthaf**
This was found by @liam-o-marsh in #240, and I cherry-picked it to merge the fix quicker. 

Ping @rosecers who wants to use this!

----
:books: Documentation preview :books:: https://rascaline--260.org.readthedocs.build/en/260/

### CODE REVIEW

**Comment by rosecers** [comment]
Much appreciated on my end! Code looks great to this noob -- let me know if you want a formal review from me.

**Comment by Luthaf** [comment]
I think the NeighborsList calculator is now wrong with these changes, around here: https://github.com/Luthaf/rascaline/blob/2851d47ab3471057f3794ac2b48b0417add2f748/rascaline/src/calculators/neighbor_list.rs#L568-L571

The condition should account for the cell shift as well.

File: python/rascaline/rascaline/systems/ase.py [hunk 1]
```diff
@@ -91,10 +91,23 @@ def compute_neighbors(self, cutoff):
...
         nl_result = neighborlist.neighbor_list("ijdDS", self._atoms, cutoff)
         for i, j, d, D, S in zip(*nl_result):
+            # we want a half neighbor list, so drop all duplicated neighbors
             if j < i:
-                # we want a half neighbor list, so drop all duplicated
-                # neighbors
                 continue
+            elif i == j:
+                if S[0] == 0 and S[1] == 0 and S[2] == 0:
+                    # only create pair with the same atom twice if the pair spans more
```
Comment (PicoCentauri) [inline python/rascaline/rascaline/systems/ase.py:99]:
```suggestion
                    # only create pairs with the same atom twice if the pair spans more
```

File: python/rascaline/tests/systems/ase.py [hunk 2]
```diff
@@ -145,3 +146,54 @@ def test_no_pbc_cell():
...
+            (2.19266, 2.11524, 3.86858),
+            (2.52936, 4.62777, 3.87771),
+            (2.01817, 0.85408, 3.87087),
+        ],
+        cell=[[3.57941, 0, 0], [0.682558, 5.02733, 0], [0.285565, 0.454525, 7.74858]],
+        pbc=True,
+    )
+
+    calculator = SphericalExpansion(
+        cutoff=9,
```
Comment (PicoCentauri) [inline python/rascaline/tests/systems/ase.py:168]:
```suggestion
        cutoff=9,  # choose a cutoff that is larger then the cell to test for "ghost" atoms
```

File: python/rascaline/tests/systems/ase.py [hunk 3]
```diff
@@ -145,3 +146,54 @@ def test_no_pbc_cell():
     )
     with pytest.warns(Warning, match=message):
         AseSystem(atoms)
+
+
+def test_same_spherical_expansion():
+    system = ase.Atoms(
```
Comment (PicoCentauri) [inline python/rascaline/tests/systems/ase.py:152]: I have the feeling that this cell is this to small to check for more then one unit cell?

File: rascaline/src/systems/neighbors.rs [hunk 4]
```diff
@@ -250,10 +250,28 @@ impl CellList {
...
+                                        continue;
+                                    }
+
+                                    if (shift[0] + shift[1] + shift[2] < 0) ||
+                                        (
+                                            (shift[0] + shift[1] + shift[2] == 0) &&
+                                            (shift[2] < 0 || (shift[2] == 0 && shift[1] < 0))
+                                        )
+                                    {
+                                        // When creating pairs between an atom
```
Comment (PicoCentauri) [inline rascaline/src/systems/neighbors.rs:266]: I understand what is going on here but maybe a picture would help. But I agree and accept, if this is too much work.

**Review by PicoCentauri — COMMENTED** [review]
Good to have this fixed.

File: python/rascaline/tests/systems/ase.py (hunk 3 above)
Comment (Luthaf) [reply python/rascaline/tests/systems/ase.py:152]: It's 3.6 in one direction, you should get pairs with a shift of 2 with a cutoff of 9

File: rascaline/src/systems/neighbors.rs (hunk 4 above)
Comment (liam-o-marsh) [reply rascaline/src/systems/neighbors.rs:266]: does this work?

```
──────┼─────┼─────┼─────┼─────┼
     ╭╯    ╭╯    ╭╯    ╭╯    ╭╯
     │     │     │   × │     │
     │     │     │  ╱  │     │
    ╭╯    ╭╯    ╭╯ ╱  <- first pair
────┼─────┼─────┼─────┼─────┼
   ╭╯    ╭╯   ╱╭╯    ╭╯    ╭╯
   │     │   × │     │     │
   │     │  ╱  │     │     │
  ╭╯    ╭╯ ╱  <- second pair, redundant
──┼─────┼─────┼─────┼─────┼
 ╭╯    ╭╯    ╭╯    ╭╯    ╭╯
 │    ╱│     │     │     │
 │   × │     │     │     │
╭╯    ╭╯    ╭╯    ╭╯    ╭╯
┼─────┼─────┼─────┼─────┼
```

Comment (Luthaf) [reply rascaline/src/systems/neighbors.rs]: We looked at the code very carefully again, and I think we both understand it now! I might do a diagram of the 2D version of it, but trying to represent the 3D version in ASCII is a bit clunky

Comment (liam-o-marsh) [reply rascaline/src/systems/neighbors.rs]: I think I understand what the hard part is in the understanding. my reasoning is:
- for two same-atom pairs characterised by (h,j,k) and (-h,-j,-k), one of them has to be ignored. how do you choose?
- first try to divide space along the h+j+k=0 plane.
- if it doesn't work, divide by the k=0 plane
- if it still doesn't work, divide by the j=0 plane.
- if the two pairs cannot be separated by any of those planes, then they are both at (h=0,j=0,k=0), which isn't possible

Comment (Luthaf) [reply rascaline/src/systems/neighbors.rs]: Yes, that was the issue. I split the condition in two and added a small illustration of the 2D case

**Review by PicoCentauri — APPROVED** [review]

**Pull request merged**